File: sdk/src/asset_sdk/types/action.py
"""Pydantic models for Action resources (mcap → seg → action)."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ActionSourceType(str, Enum):
    human = "human"
    algo = "algo"
    rule = "rule"
    system = "system"


class Action(BaseModel):
    action_id: str
    asset_id: str
    start_ns: int = Field(description="half-open interval start (ns)")
    end_ns: int = Field(description="half-open interval end (ns)")
    action_index: Optional[int] = None
    primary_label: Optional[str] = None
    labels: list[str] = Field(default_factory=list)
    description: Optional[str] = None
    attrs: dict[str, Any] = Field(default_factory=dict)
    source_type: ActionSourceType = ActionSourceType.human
    source_name: Optional[str] = None
    source_version: Optional[str] = None
    run_id: Optional[str] = None
    confidence: Optional[float] = None
    external_id: Optional[str] = None
    tenant_id: Optional[str] = None
    project_id: Optional[str] = None
    is_deleted: bool = False
    version: int = 1
    created_at: datetime
    updated_at: datetime


class ActionCreate(BaseModel):
    start_ns: int
    end_ns: int
    action_index: Optional[int] = None
    primary_label: Optional[str] = None
    labels: list[str] = Field(default_factory=list)
    description: Optional[str] = None
    attrs: dict[str, Any] = Field(default_factory=dict)
    source_type: ActionSourceType = ActionSourceType.human
    source_name: Optional[str] = None
    source_version: Optional[str] = None
    run_id: Optional[str] = None
    confidence: Optional[float] = None
    external_id: Optional[str] = None


class ActionList(BaseModel):
    items: list[Action] = Field(default_factory=list)
    asset_id: str
    total: int = 0

File: backend/internal/models/action.go
package models

import "time"

// Allowed action source_type values. Mirrors actions.source_type CHECK semantics
// (validated at the usecase boundary).
const (
	ActionSourceHuman  = "human"
	ActionSourceAlgo   = "algo"
	ActionSourceRule   = "rule"
	ActionSourceSystem = "system"
)

// Action is one row in the `actions` table — a time-bounded annotation inside a
// segment asset (`mcap → seg → action`, the third layer of the business model).
//
// Invariants the persistence layer enforces:
//   - asset_id MUST reference an asset with asset_type='segment'
//   - end_ns >= start_ns (DB CHECK)
//   - actions do NOT participate in lifecycle_state / deliveries / asset list
type Action struct {
	ActionID string `json:"action_id"`
	AssetID  string `json:"asset_id"`

	StartNs     int64 `json:"start_ns"`
	EndNs       int64 `json:"end_ns"`
	ActionIndex *int  `json:"action_index,omitempty"`

	PrimaryLabel string                 `json:"primary_label,omitempty"`
	Labels       []string               `json:"labels"`
	Description  string                 `json:"description,omitempty"`
	Attrs        map[string]interface{} `json:"attrs"`

	SourceType    string   `json:"source_type"`
	SourceName    string   `json:"source_name,omitempty"`
	SourceVersion string   `json:"source_version,omitempty"`
	RunID         string   `json:"run_id,omitempty"`
	Confidence    *float64 `json:"confidence,omitempty"`
	ExternalID    string   `json:"external_id,omitempty"`
	TaskID        string   `json:"task_id,omitempty"`

	TenantID  string `json:"tenant_id,omitempty"`
	ProjectID string `json:"project_id,omitempty"`

	IsDeleted bool      `json:"is_deleted"`
	Version   int64     `json:"version"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// IsValidActionSourceType reports whether s is one of the allowed source_type
// enum values.
func IsValidActionSourceType(s string) bool {
	switch s {
	case ActionSourceHuman, ActionSourceAlgo, ActionSourceRule, ActionSourceSystem:
		return true
	}
	return false
}
